src/types.ts

```diff
@@ -2,20 +2,58 @@ import { ValueOf } from "type-fest";
 
 type DistributeType<T> = T extends any ? { type: T } : never;
 
-export type SzNumber = { type: "number"; min?: number; max?: number };
+export type SzNumber = {
+  type: "number";
+  min?: number;
+  max?: number;
+  multipleOf?: number;
+  int?: true;
+  finite?: true;
+};
+export type SzBigInt = {
+  type: "bigInt";
+  min?: bigint;
+  max?: bigint;
+  multipleOf?: bigint;
+};
+export type SzString = {
+  type: "string";
+  min?: number;
+  max?: number;
+  length?: number;
+  startsWith?: string;
+  endsWith?: string;
+  includes?: string;
+  position?: number;
+  regex?: string;
+  ip?: true;
+  version?: "v4" | "v6";
+  datetime?: true;
+  offset?: true;
+  precision?: number;
+  email?: true;
+  url?: true;
+  emoji?: true;
+  uuid?: true;
+  cuid?: true;
+  cuid2?: true;
+  ulid?: true;
```

Comment:
There is a best-practice in data modeling called sth like "making illegal state unrepresentable". The idea is that you can't express things in your program which can't exist. This type definition here goes against that, where the types allow e.g. `{ type: "string", url: true, emoji: true }` to exist, while these should be mutually exclusive.

I think this is probably where we want a union type, sth like:
```ts
export type SzString = {
  type: "string",
  min?: number,
  max?: number,
  length?: number,
  startsWith?: string,
  endsWith?: string,
} & (
  | {}
  | {
      includes: string,
      position: number,
    }
) &
  (
    | {}
    | { stringType: "ip", version: "v4" | "v6" }
    | { stringType: "datetime", offset?: number, precision?: number }
    | { stringType: "url" | "emoji" ... }
  );
```

That way the types also inform users about which properties belong together. E.g. from the current types, one would not know that `includes` and `position` go together, apart from order. These new types express (and enforce) "you can only have one with the other"

Reply:
Yeah, makes sense, thanks. I've amended. A slight difference is that I used `object` over `{}` per typescript-eslint (warning the former could include non-objects). I have also since made `flags` optional, esp. since some might be writing this format originally in JSON.